### PR TITLE
feat(launch): add --auto-retry for unbounded bad-node failover (PR 3/3)

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

**PR 3 of 3** in the failover integration plan (PR 1 = #142 scrape, PR 2 = #143 bash lib). This PR adds `ezpz launch --auto-retry` — the ergonomic UX wrapper that the failover-integration project memory called out as the final piece.

The user-facing shape (verbatim from the brief):

```bash
# 522 nodes allocated; train on 512, reserve the rest as spares.
# Loop until success, walltime, or spare exhaustion.
ezpz launch --auto-retry --timeout 600 --np 512 -- python3 -m ezpz.examples.test
```

### Five commits, each atomic

1. **`feat(launch): add launch_autoretry module with classifier + loop`** — new `src/ezpz/launch_autoretry.py`. Pure-Python implementation of the failover loop. No bash lib dependency (the classifier is too complex to test through shell). 690 LOC.

2. **`feat(cli): add --auto-retry / --spare-nodes / --max-failover-retries flags`** — argparse surface only. Bisectable: this commit alone is a no-op on behavior. Reuses `_non_negative_int` validator from #136; adds new `_spare_nodes_value` for the `"auto"` literal.

3. **`feat(launch): wire --auto-retry through launch() and parse_args()`** — the wire-up. Cross-flag validation (mutex with `--retries`, requires explicit `--nproc`) lives in `parse_args` since argparse mutex groups can't express "non-zero `--retries`" vs the implicit-zero default. When `--auto-retry` is set, splits the PBS nodelist into active + spare, persists active to `logs/failover-<jobid>/active.hostfile`, points the launcher at that path, and dispatches through `run_with_auto_retry` instead of `_run_with_retries`.

4. **`test(launch): cover --auto-retry classifier, allocation, and loop`** — 58 tests across 5 classes. One per row of the termination matrix.

5. **`docs(launch): document --auto-retry, --spare-nodes, --max-failover-retries`** — new section in `docs/cli/launch/index.md` with the termination matrix as a markdown table.

## The termination matrix

The classifier (`launch_autoretry.classify_attempt`) is a pure function over `(rc, log_path, scraped_bad_nodes, prior_had_progress, has_spares)`. Nine outcomes:

| # | Condition                                              | Result                |
|---|--------------------------------------------------------|-----------------------|
| 1 | exit 0 (clean inner trailer, no crash patterns)        | **SUCCESS** → return 0 |
| 2 | exit 143 (walltime SIGTERM), no crash patterns         | **WALLTIME** → return rc |
| 3 | exit 143 *with* crash patterns in log                  | bad-node retry (real failure raced the walltime kill) |
| 4 | exit 124 (idle-output watchdog tripped)                | bad-node retry (silent hang → blind rotation) |
| 5 | any other non-zero, scraper found named host(s)        | swap_in named → retry |
| 6 | any other non-zero, scraper found nothing              | swap_one_blind → retry |
| 7 | two consecutive attempts with zero `step=` markers     | **STUCK_PRE_TRAINING** → return rc |
| 8 | bad-node verdict but no spares left                    | **EXHAUSTED** → return rc |
| 9 | SIGINT (Ctrl-C)                                        | **INTERRUPTED** → return 130 |

Every termination logs `FAILOVER STOP: <reason>` for post-mortem grep.

The #7 guard (two attempts with zero `step=` markers) replaces a numeric "max consecutive blind rotations" cap. The intent is to catch broken configs / missing datasets / pre-training-loop bugs *before* they burn the entire spare pool — if two attempts in a row die before `History.update` prints its first `step=N` line, no amount of node-swapping will help.

## Design decisions (preserving user intent)

- **`--auto-retry` and `--retries` are separate flags, not merged.** Two distinct concepts: `--retries N` is bounded process-level retry on the *same* nodes. `--auto-retry` is unbounded node-level failover. They're mutually exclusive.
- **`--auto-retry` requires explicit `--nproc`.** We refuse to guess the active-host count. Errors at parse time with a clear message.
- **`--spare-nodes` defaults to `"auto"` (= `total_pbs - $nproc`).** Pass an int for explicit cap; pass nothing and you get the whole unused allocation.
- **Default `--timeout` becomes 1800s** when `--auto-retry` is set (matches `FAILOVER_IDLE_TIMEOUT` in `failover.sh`). Catches silent xccl hangs before they burn full walltime.
- **Pure Python, no bash-lib dependency.** The classifier has too many edge cases to test through shell. The two paths (`--auto-retry` and `src/ezpz/bin/failover.sh`) share the scraper (`ezpz.failover.scrape_bad_nodes`) but have independent retry/swap mechanics. ~25 lines of `sed` in the bash lib isn't worth sharing.

## Hard-won crash-detection logic (cross-checked with `failover.sh`)

The same regex set lives in both places (`src/ezpz/launch_autoretry.py:_CRASH_PATTERNS_RX` and `src/ezpz/bin/failover.sh`). If you add a pattern to one, add it to the other.

- **ANSI-stripping before parsing `Execution finished with N`** — the trailer is colored; naive regex pulls `1` out of `[1;36m` instead of the real exit code (e.g. 127). Test: `TestPureHelpers.test_extract_inner_rc_ansi`.
- **Crash-pattern override on shell exit 0** — the wrapper sometimes exits clean even when the mpiexec child crashed. If the log shows OOM / gloo / shepherd-9 / EOFError signatures, treat as failure.
- **Crash-pattern override on exit 143** — real bad-node failures can surface as walltime SIGTERM when mpiexec teardown races the wallclock kill.

## Tests

`tests/test_launch_autoretry.py` — 58 tests across 5 classes:

- **TestPureHelpers (14)** — ANSI stripping, `inner_rc` extraction (incl. the colored-trailer regression), crash patterns, progress markers, spare-count derivation, backoff progression.
- **TestClassifyAttempt (14)** — one test per row of the termination matrix above, plus wrapper-lied-success override, crash-on-clean-exit override, missing-log safety.
- **TestNodeAllocation (8)** — split semantics, `active.hostfile` + `bad_nodes.txt` round-trip, `swap_in` skip-unknown / multi-host / out-of-spares handling, `swap_one_blind` rotation order, `has_spares` property.
- **TestRunWithAutoRetry (8)** — the loop with `_FakeRunner` substituted for `_run_attempt_with_tee`. Covers success-first-attempt, recovery-after-blind, named-host-swap-then-success, walltime-no-retry, the STUCK_PRE_TRAINING guard, EXHAUSTED-when-spares-drain, max_failover_retries cap, progress-clearing-the-guard, and unbounded-default exhausting all spares.
- **TestArgparseValidation (9)** — cross-flag mutex, required-explicit-nproc, `--spare-nodes` accepts `"auto"` literal + int, rejects `"AUTO"` (case-sensitive — same lesson as the scraper's `_resolve_machine_key`) and negatives, back-compat for existing `--retries` / `--timeout` without `--auto-retry`.

## Results

- **58/58 new tests pass** locally
- **131/131** in the combined launch + watchdog + scrape suite (no regressions)

## Follow-up

- [ ] On a real Aurora allocation: run `ezpz launch --auto-retry --np <N>` against a known-flaky job, verify the swap-and-retry actually swaps and retries (the bash lib has the same need from PR #143; both can be validated in one Aurora session)
- [ ] Consider adding an `ezpz launch --auto-retry --dry-run` mode that prints the resolved (active, spare, timeout) plan and exits, so users can sanity-check the allocation split before burning compute

## Test plan

- [x] `pytest tests/test_launch_autoretry.py -q` → 58/58 PASS
- [x] `pytest tests/test_launch_autoretry.py tests/test_launch_watchdog.py tests/test_launch.py tests/test_failover_scrape.py -q` → 131/131 PASS
- [ ] On Aurora: `ezpz launch --auto-retry --np <N> --timeout 600 -- python3 -m ezpz.examples.test`, kill one of the assigned nodes mid-run, verify the loop swaps and continues

## Summary by Sourcery

Introduce an auto-retry failover mode to `ezpz launch` that automatically reallocates work from bad nodes using spare nodes and a classification-driven retry loop.

New Features:
- Add `--auto-retry`, `--spare-nodes`, and `--max-failover-retries` flags to `ezpz launch` for automatic bad-node failover with configurable spare capacity and retry bounds.
- Implement a Python-based auto-retry engine that classifies attempt outcomes from logs and exit codes to drive node swapping and retries, including idle-output watchdog handling.

Enhancements:
- Wire the new auto-retry engine into the launch path so host allocation is split into active and spare sets before topology inference, with per-job failover logs and hostfiles written under `logs/failover-<jobid>`.

Documentation:
- Extend the `ezpz launch` CLI documentation with a dedicated section for `--auto-retry`, describing required flags, spare-node policy, default timeouts, termination matrix, and generated log artifacts.

Tests:
- Add a comprehensive test suite for the auto-retry logic, covering helper functions, the termination classifier, node allocation behavior, the retry loop, and argparse validation for the new flags.

### CODE REVIEW

**Comment by saforem2** [comment]
Thanks for the careful review — all 9 actionable items addressed across 6 commits (`7b1a80f` through `ddf7713`).

## P0 bugs (real correctness issues)

**Codex P1 — `--hostfile` was silently ignored under `--auto-retry`** (`7b1a80f`)
The auto-retry path always pulled from the scheduler nodelist, discarding any explicit `--hostfile`. A user pre-filtering their hostfile (e.g. dropping known-bad nodes from a previous job) would unknowingly get those bad nodes back. Fix: new `_resolve_auto_retry_node_pool` helper that prioritizes `--hostfile` over the scheduler list, with clear errors when both sources are empty. 8 new tests pin the priority + whitespace handling + error paths.

**Codex P2 — retry cap checked AFTER backoff sleep** (`ac9b6d2`)
With `--max-failover-retries 0` the loop slept for 5s before checking whether it was exhausted. Even non-zero caps wasted a backoff sleep on the way out. Fix: reorder cap check before sleep. Two regression tests pin the "no wasted backoff before cap-exit" contract.

**Copilot — tee reader race truncating logs** (`5c59233`)
`_run_attempt_with_tee` returned on `proc.poll()` with only a 2s best-effort drain in `finally`. The classifier could see a truncated log when the child wrote a final burst right before exit. Fix: explicit `_drain_remaining()` with 30s upper bound BEFORE returning (safe because the child has already exited; reader exits naturally on stdout EOF). 1000-line burst test pins the race.

## P1 — docs/help drift on `--spare-nodes` formula (`5c9923d`)

**Copilot ×4.** The default-spare formula was documented as `total_pbs_nodes - $nproc` in 4 places, but the code uses `total_pbs_nodes - ceil($nproc / $ppn)` — accounting for the fact that `--nproc` counts ranks, not hosts. Especially misleading in the 522-node example: with Aurora's 12 ranks/node, `--np 512` is 43 hosts, not 512. Fixed all 4 spots, rewrote the example with realistic Aurora math, renamed the misleading `gpus_per_node` local to `ranks_per_node`.

## P2 — code cleanup

**Sourcery — structured classifier result, single log read** (`5d7de63`)
`classify_attempt` now returns a small frozen `ClassificationResult(reason, has_progress)` instead of a bare `TerminationReason`. The loop threads `has_progress` to the next iteration without re-reading the log. Also documented in the enum that `INTERRUPTED` is loop-produced, not classifier-produced (the SIGINT handler re-raises before reaching `classify_attempt`). Kept `INTERRUPTED` in the enum since the docs matrix row 9 references it by name.

**Copilot — drop dead `AutoRetryConfig.hostfile` field** (`ddf7713`)
Field was documented as "patched in place" but never read. The real contract: `cmd` already contains `--hostfile=<path>` resolved at `build_executable` time; `NodeAllocation` mutates that file's contents between attempts. Updated docstrings to spell out the real invariant.

**Sourcery — differentiate the two `--nproc` checks** (`ddf7713`)
`parse_args` raises `SystemExit` with the user-facing CLI message; `launch()` now raises `ValueError` with a library-level message that points CLI users at the right flag. Same invariant, different audience — no longer two duplicate `SystemExit` strings that could drift.

**Sourcery security — Popen with dynamic argv** (`ddf7713`)
False positive: `Popen` + list argv + `shell=False` is safe by construction, and `cmd` originates from argparse `REMAINDER` (the user's own shell input — equivalent to typing the command into a terminal). Added `# noqa: S603` with a comment explaining why.

## On the "xccl" question

Not a typo — that's the Intel oneCCL XPU backend on Aurora (analogous to NCCL on NVIDIA). It's the relevant library for the silent-hang case the watchdog was built around: `xccl` ignores `train_timeout_seconds` and a hung collective consumes the full PBS walltime. Keeping the existing spelling.

## Tests

- 99 → 136 tests across launch + watchdog + autoretry suites (37 new)
- All 136 pass locally
- No regressions in the existing scrape (36) or failover-lib bash (21) suites

File: tests/test_launch_autoretry.py [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+        assert rc == 1
+        assert len(runner.calls) == 4
+
+
+# ---------------------------------------------------------------------------
+# Argparse: cross-flag validation
+# ---------------------------------------------------------------------------
+
+
+class TestArgparseValidation:
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_launch_autoretry.py:832]: Addressed in `404bd23` and expanded further in `7b1a80f` — `TestResolveAutoRetryAllocation` (9 tests) covers the `"auto"`/`None`/explicit-int branches, the `spare_nodes > available` SystemExit, the `nproc > total` SystemExit, single-host edge case, `--spare-nodes 0`, and the truncation behavior. `TestResolveAutoRetryNodePool` (8 more tests) covers the `--hostfile` priority added in `7b1a80f`. Net: 17 tests on the resolver paths.

File: docs/cli/launch/index.md [hunk 2]
```diff
@@ -141,6 +150,115 @@ If every attempt fails, the final attempt's exit code is returned.
...
+datasets / pre-training-loop bugs before they burn the entire spare
+pool — if two attempts in a row die before `History.update` prints
+its first `step=N` line, no amount of node-swapping will help.
+
+### Default idle-output watchdog
+
+When `--auto-retry` is set, `--timeout` defaults to **1800 seconds**
+(30 minutes) instead of being off. This matches the
+`FAILOVER_IDLE_TIMEOUT` default in `src/ezpz/bin/failover.sh` and
+prevents silent xccl hangs from burning the full PBS walltime
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/cli/launch/index.md:354]: Not a typo — `xccl` is Intel's oneCCL XPU collective backend on Aurora (analogous to NCCL on NVIDIA). The motivating incident was an xccl-specific silent hang where `train_timeout_seconds` was ignored and the job consumed full walltime. Keeping the existing spelling.

File: src/ezpz/launch_autoretry.py [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+    # period (`Execution finished with 0.`) which would otherwise
+    # poison int().
+    tail = tail.rstrip(".,;:")
+    try:
+        return int(tail)
+    except ValueError:
+        return None
+
+
+def classify_attempt(
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch_autoretry.py:368]: Done in `5d7de63`. `classify_attempt` now returns a frozen `ClassificationResult(reason, has_progress)`; the loop reads `result.has_progress` and threads it through to the next iteration without re-reading the log. Also clarified in the enum docstring that `INTERRUPTED` is loop-produced (in the SIGINT handler) — the classifier never sees that path because `KeyboardInterrupt` is re-raised before reaching it. Kept `INTERRUPTED` in the enum since the docs matrix row 9 references it by name.

File: src/ezpz/launch_autoretry.py [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+    child_env.setdefault("PYTHONUNBUFFERED", "1")
+
+    proc = subprocess.Popen(
+        list(cmd),
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
+        bufsize=1,
+        env=child_env,
+    )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch_autoretry.py]: False positive — addressed in `ddf7713` with a `# noqa: S603` and an inline safety note. `Popen` here gets `shell=False` (the default) and a list argv from `argparse.REMAINDER`, i.e. the user's own shell input. Equivalent to typing the command into a terminal; no privilege boundary crossed.

File: src/ezpz/launch.py [hunk 5]
```diff
@@ -651,6 +751,43 @@ def launch(
...
+        gpus_per_node = ngpu_per_host or ezpz.get_gpus_per_node() or 1
+        # Round-up division: 12 ranks on 4-gpu nodes → 3 nodes.
+        nhosts_active = (ngpus + gpus_per_node - 1) // gpus_per_node
+        autoretry_log_dir = _auto_retry_log_dir(jobid)
+        autoretry_allocation, autoretry_hostfile = (
+            _resolve_auto_retry_allocation(
+                nodelist,
+                nhosts_active,
+                spare_nodes,
+                autoretry_log_dir,
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch.py:845]: Real bug, good catch. Fixed in `7b1a80f`. New `_resolve_auto_retry_node_pool` helper now prioritizes the user's `--hostfile` over the scheduler `nodelist`. Pre-filtered hostfiles (e.g. with known-bad nodes already dropped) are honored. Errors clearly when both sources are empty. 8 new tests pin the priority and the error paths.

File: src/ezpz/launch_autoretry.py [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+            backoff = _backoff_for_attempt(attempt - 1)
+            logger.warning(
+                "[auto-retry] attempt %d (prior rc=%d, sleeping %.0fs)...",
+                attempt,
+                last_rc,
+                backoff,
+            )
+            time.sleep(backoff)
+
+        if (
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch_autoretry.py]: Fixed in `ac9b6d2`. Cap check now fires BEFORE the backoff sleep. Two regression tests in `TestRunWithAutoRetry`: `test_max_failover_retries_zero_no_retry` proves the cap=0 case skips the sleep entirely, `test_max_failover_retries_one_sleeps_once` proves cap=1 sleeps exactly once (before attempt 2) and not a second time on the way to the cap-exit.

File: src/ezpz/launch_autoretry.py [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+            reader_done.set()
+
+    reader = threading.Thread(target=_drain, daemon=True)
+    reader.start()
+
+    try:
+        while True:
+            rc = proc.poll()
+            if rc is not None:
+                return rc
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch_autoretry.py:556]: Fixed in `5c59233`. `_run_attempt_with_tee` now does an explicit `_drain_remaining()` (30s upper bound) BEFORE returning at both exit points (normal child exit + watchdog kill). New `test_full_log_captured_after_child_exit` writes a 1000-line burst immediately before exit and asserts every line including the last makes it to disk. NOT wired on the SIGINT path — the classifier never reads the log there.

File: src/ezpz/launch_autoretry.py [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+
+    cmd: list[str]
+    """Full launcher command line (mpiexec + user command)."""
+
+    hostfile: Path
+    """Path to the active hostfile. Re-read at every attempt — the
+    same path, with its contents mutated by ``NodeAllocation.swap_in``
+    / ``swap_one_blind``. Pointing the launcher at this path means
+    the launcher always sees the current active set."""
+
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch_autoretry.py]: Dropped in `ddf7713`. The field was documentation-only — `run_with_auto_retry` never read it. Updated `AutoRetryConfig` docstrings to spell out the real contract: `cmd` already contains `--hostfile=<path>` resolved at `build_executable` time; `NodeAllocation` mutates that file's contents in place between attempts.

File: src/ezpz/cli/flags.py [hunk 9]
```diff
@@ -391,6 +405,48 @@ def build_launch_parser(
...
+        "--spare-nodes",
+        type=_spare_nodes_value,
+        default=None,
+        dest="spare_nodes",
+        help=(
+            "Number of spare nodes to reserve for --auto-retry "
+            "swap-ins. Pass 'auto' (the default when --auto-retry is "
+            "set) to derive from total_pbs_nodes - $nproc. Pass an "
+            "integer for an explicit count. Ignored when --auto-retry "
+            "is not set."
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/cli/flags.py]: Fixed in `5c9923d` — help text now reads `total_pbs_nodes - ceil($nproc / $ppn)` with a parenthetical "ranks → hosts, since --nproc counts ranks". Same fix landed in the `_spare_nodes_value` docstring.

File: docs/cli/launch/index.md [hunk 10]
```diff
@@ -85,6 +85,15 @@ allocated to your job.
         --timeout IDLE_TIMEOUT_S
                                 Idle-output watchdog timeout in seconds. Off by default.
         --retries RETRIES     Re-execute on non-zero exit, up to N times. Default: 0.
+        --auto-retry          Unbounded bad-node failover loop. Mutually
+                                exclusive with --retries. Requires explicit --nproc.
+        --spare-nodes SPARE_NODES
+                                Spare-node pool for --auto-retry. "auto" (default)
+                                derives from total_pbs_nodes - $nproc; pass an int
+                                for an explicit cap.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/cli/launch/index.md]: Fixed in `5c9923d` — synopsis help-extract corrected to match `cli/flags.py` (`total_pbs_nodes - ceil($nproc / $ppn)`).

File: src/ezpz/launch.py [hunk 11]
```diff
@@ -651,6 +751,43 @@ def launch(
...
+                "to split the PBS allocation into active + spare."
+            )
+        if nodelist is None or not nodelist:
+            raise SystemExit(
+                "--auto-retry: failed to read the PBS nodelist (no "
+                "active job?). Cannot split into active + spare."
+            )
+        gpus_per_node = ngpu_per_host or ezpz.get_gpus_per_node() or 1
+        # Round-up division: 12 ranks on 4-gpu nodes → 3 nodes.
+        nhosts_active = (ngpus + gpus_per_node - 1) // gpus_per_node
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch.py]: Renamed in `5c9923d`. The local `gpus_per_node` is now `ranks_per_node` with a comment explaining the `--ppn` (== `ngpu_per_host`) → ranks-per-node mapping and the `or get_gpus_per_node() or 1` fallback. Also extracted the round-up division into `_ranks_to_hosts(nranks, ranks_per_host)` so the formula has one canonical home (`TestRanksToHosts` x5 tests).

File: docs/cli/launch/index.md [hunk 12]
```diff
@@ -141,6 +150,115 @@ If every attempt fails, the final attempt's exit code is returned.
...
+```
+
+### Spare-node policy (`--spare-nodes`)
+
+By default (`--spare-nodes auto`), the spare pool is
+`total_pbs_nodes - $nproc/$ppn`. So if PBS gave you 522 nodes and
+you ask for 512 training ranks on 12-GPU nodes, the spare pool is
+`522 - 43 = 479` — though you'll usually want to reserve fewer in
+practice (a node typically isn't worth using if 50 others failed
+on it first).
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/cli/launch/index.md]: Fixed in `5c9923d`. "Spare-node policy" section rewritten with correct math and a paragraph explaining the ranks-vs-hosts conversion. Worked example: 50 nodes on Aurora (12 ranks/node) + `--np 512` → `ceil(512/12) = 43` active hosts, 7 spare.

File: docs/cli/launch/index.md [hunk 13]
```diff
@@ -141,6 +150,115 @@ If every attempt fails, the final attempt's exit code is returned.
...
+ezpz scrapes the log for known bad-node signatures (Aurora PALS
+shepherd-9, gloo TCP peer-closed), swaps each named host out for a
+spare from the rest of the PBS allocation, and re-runs the command.
+Unlike `--retries N`, the loop is *unbounded* by default — it
+continues until one of the conditions in the **termination matrix**
+below fires.
+
+```bash
+# 522 nodes allocated by PBS; use 512 for training, reserve 10 as
+# spares. Loop until success / walltime / spare exhaustion.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/cli/launch/index.md]: Fixed in `5c9923d`. The original example treated 512 as a host count (which was wrong — `--np` is ranks). Rewrote with a realistic 50-node Aurora allocation, explicit ceiling-divide arithmetic in a comment, and the correct spare count (7).
